tools/chpl-language-server/src/shimchapel.py

```diff
@@ -0,0 +1,122 @@
+    ctx.set_module_paths(module_paths, files)
+
+    for file in files:
+        asts = ctx.parse(file)
+
+        # Trigger scope resolution to make sure other files are parsed.
+        # XXX: hack
+        for ast, _ in each_matching(asts, Identifier):
+            ast.to_node()
+
+    return [
+        parsed_path
+        for parsed_path in ctx.introspect_parsed_files()
+        if not ctx.is_bundled_path(parsed_path)
+    ]
+
+
+if __name__ == "__main__":
+    if sys.argv[0].endswith("shimchapel.py"):
+        # Top-level invocation; behave as the driver script.
+
+        # Create a temporary directory to add the 'chpl' wrapper script
+        tempdir = tempfile.TemporaryDirectory()
+        wrapper = os.path.join(tempdir.name, "chpl")
+        os.symlink(__file__, wrapper)
+
+        # Modify path to include the wrapper script at the front
+        chpl = subprocess.check_output(["which", "chpl"]).decode(sys.stdout.encoding).strip()
+        newpath = os.pathsep.join([tempdir.name, os.environ["PATH"]])
+        os.environ["CHPL_SHIM_TARGET_PATH"] = tempdir.name
+        os.environ["CHPL_SHIM_REAL_COMPILER_PATH"] = chpl
+        os.environ["PATH"] = newpath
+
+        # Invoke the script I was asked to invoke
+        subprocess.run(sys.argv[1:])
+
+        # At this point, our subprocessses have finished and created flags
+        # in the tempdir. Iterate all the files in the dir and merge the results.
+        commands = defaultdict(list)
+
+        for file in os.listdir(tempdir.name):
+            file = os.path.join(tempdir.name, file)
+            if os.path.isfile(file) and file.endswith(".json"):
+                with open(file) as f:
+                    from_file = json.load(f)
+                    for key, value in from_file.items():
+                        commands[key].append(value)
+
+        with open(".cls-info.json", "w") as f:
+            json.dump(commands, f)
+
+    else:
+        # Being invoked a wrapper script
+        real_compiler = os.environ["CHPL_SHIM_REAL_COMPILER_PATH"]
+        tmpfile_path = os.environ["CHPL_SHIM_TARGET_PATH"]
+
+        parser = argparse.ArgumentParser()
+        parser.add_argument('filename', nargs='*')
+        parser.add_argument('-M', '--module-dir', action='append', dest="module_dirs", default=[])
+
+        args, unknown = parser.parse_known_args()
+        files = args.filename
+
+        # Since we don't have a perfectly matching arg-parser, iterate argv again
+        # to look for .chpl files
+        for arg in sys.argv:
+            if arg.endswith(".chpl"):
+                files.append(arg)
+
+        files = [os.path.abspath(file) for file in files if file.endswith(".chpl")]
+        files += list_parsed_files(files, args.module_dirs)
+        files = list(set(files))
+
+        commands = {}
+        for file in files:
+            invocation = " ".join(sys.argv).replace(__file__, real_compiler)
+            commands[file] = {
+            #   "invocation": invocation,
```

Comment:
I left it commented out because right now it doesn't seem to be useful, and it really balloons the file size. On the other hand, it seems like it could be useful.

Reply:
I agree, it seems like that may eventually be very useful, but I understand the file size concerns. I would also then comment out `invocation =`, and put a note as to why its commented out